### PR TITLE
roles/srv-announcer: simplify fetching

### PR DESCRIPTION
**Pull request by flokli**
There's few value in providing three different ways to retrieve the srv
announcer. In fact, it's debatable on whether this should belong in this
ansible role at all, given it's something very specific to our
environment (and the lack of service discovery).

Simplify this, by only accepting
```
 - srv_announcer_artifact_file_url
 - srv_announcer_artifact_checksum
```

Which needs to be set from the outside while instantiating this role.

`srv_announcer_artifact_checksum` now is also piped directly to
`get_url.checksum`, meaning it'll need to have the hashing algorithm and
a colon in front.

### CODE REVIEW

**Comment by flokli** [comment]
Existing code using this would need to prefix `sha256:` to `srv_announcer_artifact_checksum`, because this doesn't hardcode it anymore.

**Review by akshaymankar — COMMENTED** [review]
I like this but I'd wait for @lucendio to confirm this doesn't break anything in avs-test to deploy "unrealesed" artifacts.

File: roles/srv-announcer/tasks/install.yml [hunk 1]
```diff
@@ -1,47 +1,23 @@
...
-  when:
-    - srv_announcer_version is defined
-    - srv_announcer_artifact_file_path is not defined
-    - srv_announcer_artifact_file_url is not defined
-    - srv_announcer_artifact_checksum is not defined
-
 - name: downloading {{ srv_announcer_service_name }} artifact by URL
   get_url:
     url: "{{ srv_announcer_artifact_file_url }}"
-    checksum: "sha256:{{ srv_announcer_artifact_checksum }}"
```
Comment (lucendio) [inline roles/srv-announcer/tasks/install.yml:25]: question: atm the build process only spits out SHA256, so what would be the reason to hand this over to the inventory author?

File: roles/srv-announcer/tasks/install.yml [hunk 2]
```diff
@@ -1,47 +1,23 @@
...
-    dest: "{{ srv_announcer_artifact_target_path }}"
+    checksum: "{{ srv_announcer_artifact_checksum }}"
+    dest: "/tmp/srv-announcer.tar.gz"
```
Comment (lucendio) [inline roles/srv-announcer/tasks/install.yml:5]: question: what is the reasoning behind moving away from using the existing constant `srv_announcer_artifact_target_path`?

File: roles/srv-announcer/tasks/install.yml [hunk 3]
```diff
@@ -1,47 +1,23 @@
...
     - srv_announcer_artifact_checksum is defined
-    - srv_announcer_version is not defined
-    - srv_announcer_artifact_file_path is not defined
-
 
 - name: unpacking artifact
   unarchive:
     remote_src: yes
-    src: "{{ srv_announcer_artifact_target_path }}"
+    src: "/tmp/srv-announcer.tar.gz"
```
Comment (lucendio) [inline roles/srv-announcer/tasks/install.yml:13]: question: what is the reasoning behind moving away from using the existing constant `srv_announcer_artifact_target_path`?

File: roles/srv-announcer/tasks/install.yml [hunk 4]
```diff
@@ -1,47 +1,23 @@
...
     dest: '/tmp/'
     list_files: yes
   register: unpacked_srv_announcer
 
 - name: copying {{ srv_announcer_service_name }} executable to right location
   copy:
-    src: "{{ unpacked_srv_announcer.dest }}{{ unpacked_srv_announcer.files[0] }}"
+    src: "/tmp/srv-announcer.tar.gz"
```
Comment (lucendio) [inline roles/srv-announcer/tasks/install.yml:20]: issue: archive is being copied instead of the executable
suggestion: revert the change
question: what is the reasoning behind moving away from using the existing constant `srv_announcer_artifact_target_path`?

**Review by lucendio — COMMENTED** [review]
Can you elaborate on your motivation? My understanding is that the use case we currently have in staging and prod already works with the role as is. If that was what you had in mind.

I tried to explain the reasoning behind the 3 scenarios multiple times in the past, even in some of the PRs here, and I still believe they are valid, including the *only-provide-a-version* one. So, I'm opposed to this change.

If it's about the step of maintaining the versions list, I could think of setting up a github action that creates a PR on every new srv-announcer release (in case change rate on the announcer code remains high).

> [...] this doesn't break anything in avs-test to deploy "unrealesed" artifacts

No issue here, since the srv-announcer is not deployed in the SFT test machine in AVS test.

File: roles/srv-announcer/tasks/install.yml (hunk 1 above)
Comment (flokli) [reply roles/srv-announcer/tasks/install.yml:25]: `checksum` doesn't really imply `sha256`.

This is about allowing users to provide all other checksum methods supported by `get_url.checksum`.

File: roles/srv-announcer/tasks/install.yml (hunk 2 above)
Comment (flokli) [reply roles/srv-announcer/tasks/install.yml:5]: Strictly speaking, even this is wrong, and we shouldn't drop files in before-known temporary folders. See https://cwe.mitre.org/data/definitions/377.html.

We should either create a temporary folder to download and unpack the artifact (using the `tempfile` module), or just serve the (single) binary directly.

File: roles/srv-announcer/tasks/install.yml (hunk 3 above)
Comment (flokli) [reply roles/srv-announcer/tasks/install.yml:13]: Replied above - we should get rid of the unpacking phase alltogether.

File: roles/srv-announcer/tasks/install.yml (hunk 4 above)
Comment (flokli) [reply roles/srv-announcer/tasks/install.yml:20]: yes, you're right - Based on above discussion, let me know what you think about downloading the single file to `{{ srv_announcer_bin_path }}` directly.

File: roles/srv-announcer/tasks/install.yml (hunk 1 above)
Comment (lucendio) [reply roles/srv-announcer/tasks/install.yml:25]: Since we are controlling the supply chain, and thus the checksum(s), it makes little sense to me to expose the whole set of capabilities that `get_url.checksum` offers.

File: roles/srv-announcer/tasks/install.yml (hunk 4 above)
Comment (lucendio) [reply roles/srv-announcer/tasks/install.yml:20]: `tempfile` is reasonable, I agree. I didnt know about it at the time of writing. However, this wouldn't change my overall opinion on the goal of this PR.

File: roles/srv-announcer/tasks/install.yml (hunk 1 above)
Comment (arianvp) [reply roles/srv-announcer/tasks/install.yml:25]: This discussion feels a bit bike-sheddy to me.   Having the type of hash next to the shasum is good for documentation; so I'd actually prefer personally if we keep the two together (or rename the variable to contain the hashing mechanism used).

File: roles/srv-announcer/tasks/install.yml (hunk 2 above)
Comment (flokli) [reply roles/srv-announcer/tasks/install.yml:5]: Would you be fine if we'd serve the file directly?

Comment (lucendio) [reply roles/srv-announcer/tasks/install.yml:5]: I dont see a reason for not serving an archive. Being able to provide some sort of manual/information alongside the binary is useful. I'll bring up a new commit to this branch that makes use of the `tempfile` task. Feel free to revert it if it's not to your liking.

File: roles/srv-announcer/tasks/install.yml [hunk 5]
```diff
@@ -1,47 +1,23 @@
...
-    dest: "{{ srv_announcer_artifact_target_path }}"
-  when:
-    - srv_announcer_artifact_file_path is defined
-    - srv_announcer_version is not defined
-    - srv_announcer_artifact_file_url is not defined
-    - srv_announcer_artifact_checksum is not defined
-
-- name: downloading {{ srv_announcer_service_name }} artifact by version
-  get_url:
-    url: "{{ srv_announcer_download_base_url }}/{{ srv_announcer_artifact_basename }}.tar.gz"
```
Comment (arianvp) [inline roles/srv-announcer/tasks/install.yml:13]: Just a suggestion. But if we move this interpolatio-logic into `defaults` into a variable named `srv_announce_url:  {{ srv_announcer_download_base_url }}/{{ srv_announcer_artifact_basename }}.tar.gz`   doesn't that make both @lucendio and @flokli happy? as then both usecases are supported?  (Override the entire URI; or just override the version)

File: roles/srv-announcer/tasks/install.yml (hunk 2 above)
Comment (arianvp) [reply roles/srv-announcer/tasks/install.yml]: Sounds good to me

File: roles/srv-announcer/tasks/install.yml (hunk 5 above)
Comment (lucendio) [reply roles/srv-announcer/tasks/install.yml:13]: If I understand you correctly, the proposal would basically be move the [deleted lines](https://github.com/wireapp/ansible-sft/pull/25/commits/c301c45d629e5c2b17ec24df6bc39cff01662eb8#diff-4e4a7fa6b5d4044be56da310925a8eaea2eb820d51c8d3abba9e8f258d0528e4L5-L6) over to `defaults/main.yml`. For that, the [checksum list](https://github.com/wireapp/ansible-sft/pull/25/commits/c301c45d629e5c2b17ec24df6bc39cff01662eb8#diff-4e4a7fa6b5d4044be56da310925a8eaea2eb820d51c8d3abba9e8f258d0528e4L17-L21) still needs to be maintained, which is - and that would be my guess - what @ffflorian wanted to get rid of. While I personally think it would be a valuable user experience to only define the version and don't care about anything else, I can see the overhead in this.

My intention with the last commit was basically to fix this PR and move on. But I'm also happy to take a step back again and discuss the three ways of specifying the artifact that is being used.

Comment (arianvp) [reply roles/srv-announcer/tasks/install.yml:13]: I think having the hashes is fine.  Lets keep it.   If things turn out to be needed to be more override-able  we can always send a patch later. Let me approve this PR so you can merge it

**Review by arianvp — APPROVED** [review]

File: roles/srv-announcer/tasks/install.yml (hunk 5 above)
Comment (arianvp) [reply roles/srv-announcer/tasks/install.yml:13]: I thought the hashes part was kept in from the discussion. I didn't look at the diff. I think  I approved too soon now. Let me open a PR that reintroduces that part. sorry for the noise
